jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java

```diff
@@ -0,0 +1,351 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.decomposition;
+
+import org.jgrapht.Graph;
+import org.jgrapht.SlowTests;
+import org.jgrapht.alg.connectivity.ConnectivityInspector;
+import org.jgrapht.generate.BarabasiAlbertForestGenerator;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.SimpleGraph;
+import org.jgrapht.util.SupplierUtil;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Tests for {@link HeavyPathDecomposition}
+ *
+ * @author Alexandru Valeanu
+ */
+public class HeavyPathDecompositionTest {
+
+    public static int log2(int n) // returns 0 for n=0
+    {
+        int log = 0;
+        if( ( n & 0xffff0000 ) != 0 ) { n >>>= 16; log = 16; }
+        if( n >= 256 ) { n >>>= 8; log += 8; }
+        if( n >= 16  ) { n >>>= 4; log += 4; }
+        if( n >= 4   ) { n >>>= 2; log += 2; }
+        return log + ( n >>> 1 );
+    }
+
+    /*
+        Count the maximum number of distinct paths on any root-to-leaf path
+     */
+    public static <V, E> int countMaxPath(Set<V> vertexSet, HeavyPathDecomposition<V, E> decomposition){
+        List<List<V>> paths = decomposition.getPathDecomposition().getPaths();
+        Map<V, Integer> whichPath = new HashMap<>();
+
+        for (int i = 0; i < paths.size(); i++) {
+            List<V> path = paths.get(i);
+
+            for (int j = 0; j < path.size(); j++) {
+                whichPath.put(path.get(j), i);
+            }
+        }
+
+        int maxim = 0;
+        HeavyPathDecomposition<V, E>.InternalState state = decomposition.getInternalState();
+
+        for (V v: vertexSet){
+            if (whichPath.containsKey(v)){
+                int lastPath = -1;
+                int cnt = 0;
+
+                while (v != null){
+                    if (lastPath != whichPath.get(v)){
+                        lastPath = whichPath.get(v);
+                        cnt++;
+                    }
+
+                    v = state.getParent(v);
+                }
+
+                maxim = Math.max(maxim, cnt);
+            }
+        }
+
+        return maxim;
+    }
+
+    public static <V, E> boolean isValidDecomposition(Graph<V, E> graph, Set<V> roots,
```

Comment:
I think you have to test for each node, that the size of the subtree rooted at a heavy edge is larger than the sum of sizes of all subtrees rooted at light edges. The log_2 property then follows.

Reply:
I don't think that works. Take for example [1, 2, 3, 4] {1 -> 2, 1->3, 1->4}. There is only one heavy edge (1 -> 2; any of them works) and the subtree rooted 2 doesn't contain more vertices than all other three combined. Let me know if I didn't understand what you meant.

The current test works as the number of heavy edges on any node-to-root path is at most log2(|V|.)

Reply:
You are right, my definition was wrong. A heavy edge is when the child has size larger than half of the size of the parent. Light edges are the remaining ones.

However, the number of light edges is log(n) on any root-to-leaf path. This is a consequence of the above definition, which means that if you follow a light edge you can have at most half the nodes.

So your test should be to verify that each light edge has this property. The logn bound follows from that. Just do a post-order and count the number of nodes of each subtree (including the subtree root) and then verify for each light edge.

Reply:
I guess for the test to be complete, you also need to test the property of the heavy edges. Then the logn bound check will be superfluous.

Reply:
I'd like to keep that last test. If it is redundant at this stage.